core/cafe_service.go

```diff
@@ -615,6 +618,28 @@ func (h *CafeService) handleRegistration(pid peer.ID, env *pb.Envelope) (*pb.Env
 		return h.service.NewError(403, errForbidden, env.Message.RequestId)
 	}
 
+	// does the provided token match?
+	s := strings.Split(reg.Token, "+")
+	if len(s) != 2 {
+		h.service.NewError(403, errForbidden, env.Message.RequestId)
+	}
+	id, token := s[0], s[1]
+	plainBytes, err := base58.FastBase58Decoding(token)
+	if err != nil {
+		return h.service.NewError(403, errForbidden, env.Message.RequestId)
+	}
+
+	encodedToken := h.datastore.CafeDevTokens().Get(id)
+	if encodedToken == nil {
+		return h.service.NewError(403, errForbidden, env.Message.RequestId)
+	}
+
+	err = bcrypt.CompareHashAndPassword([]byte(encodedToken.Token), plainBytes)
```

Comment:
We (for safety sake, must) use `CompareHashAndPassword`, which is why we use a concated token on the client side, and index them by their ksuid component. This makes it possible to pull out the token from the db, which contains the salt and hashing params. Then we can use the constant time comparison function from bcrypt to compare the tokens.

Reply:
I think we'll want to reconsider the concatenation - makes it feel overly complicated. You can look up the token by `token`, and an index to be more efficient

Reply:
see other comments for why this isn't possible. however, have moved to different approach using random bytes for token and and base58 encoding the whole token